pep-0705.rst

```diff
@@ -0,0 +1,259 @@
+
+    from typing import TypedDict
+
+    class MovieRecord(TypedDict):
+        name: str
+        year: int
+
+    movie: MovieRecord = {'name': 'Blade Runner',
+                          'year': 1982}
+    print(movie_string(movie))
+
+Note that currently the code above will fail to type check if we define ``Movie`` based on ``TypedDict``, and there is no replacement.
+
+Motivation
+==========
+
+Representing structured data using (potentially nested) dictionaries with string keys is a common pattern in Python programs. :pep:`589` allows these values to be type checked when the exact type is known up-front, but it is hard to write read-only code code that accepts more specific variants: for instance, where fields may be subtypes or restrict a union of possible types. This is an especially common issue when writing APIs for services, which may support a wide range of input structures, and typically do not need to modify their input.
+
+Take for example the function ``movie_string`` defined above. Suppose its type signature is defined using a ``TypedDict``::
+
+    from typing import NotRequired, TypedDict
+
+    class Movie(TypedDict):
+        name: str
+        year: NotRequired[int | None]
+
+    class MovieRecord(TypedDict):
+        name: str
+        year: int
+
+    def movie_string(movie: Movie) -> str:
+        if movie.get("year") is None:
+            return movie["name"]
+        else:
+            return f'{movie["name"]} ({movie["year"]})'
+
+Attempting to pass a ``MovieRecord`` into ``movie_string`` results (using mypy) in the error::
+
+    Argument 1 to "movie_string" has incompatible type "MovieRecord"; expected "Movie"
+
+This particular use case should be type-safe, but the type checker correctly stops the
+user from passing a ``MovieRecord`` into a ``Movie`` argument in the general case because
+the ``Movie`` class has mutator methods that could potentially allow the function to break
+the type constraints in ``MovieRecord`` by setting ``movie["year"] = None`` or ``del movie["year"]``.
+The problem disappears if we don't have mutator methods in ``Movie``. This could be achieved by defining an immutable interface using a :pep:`544` Protocol::
+
+    from typing import Literal, Protocol, overload
+
+    class Movie(Protocol):
+        @overload
+        def get(self, key: Literal["name"]) -> str: ...
+
+        @overload
+        def get(self, key: Literal["year"]) -> int | None: ...
+
+        @overload
+        def __getitem__(self, key: Literal["name"]) -> str: ...
+
+        @overload
+        def __getitem__(self, key: Literal["year"]) -> int | None: ...
+
+This is very repetitive, easy to get wrong, and is still missing important method definitions like ``__contains__()`` and ``keys()``.
+
+The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and support the cases exemplified above.
```

Comment:
```suggestion
Rationale
======

The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and supports the cases exemplified above.
```

* IMO, this is really your Rationale section here, since it is explaining the benefits for your proposed _solution_, as opposed to laying out the _problem_,.
* Fix grammar (singular/plural agreement)

Also, it would be really nice to avoid overloading "type"/"types" so much to refer to three different things here, but I'm not sure exactly what to suggest.